backend/metamodel/syson-sysml-metamodel/src/main/java/org/eclipse/syson/sysml/util/LibraryNamespaceProvider.java

```diff
@@ -0,0 +1,196 @@
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.syson.sysml.Element;
+import org.eclipse.syson.sysml.LibraryPackage;
+import org.eclipse.syson.sysml.Namespace;
+import org.eclipse.syson.sysml.helper.EMFUtils;
+import org.eclipse.syson.sysml.helper.NameHelper;
+
+/**
+ * Provider of {@link Namespace} stored in {@link LibraryPackage}. It implementation offer a way to speed computation by
+ * registering immutable {@link LibraryPackage} such as SyML and KerML default libraries. The qualified name of element
+ * stored in such libraries are stored in cache.
+ *
+ * @author Arthur Daussy
+ */
+public class LibraryNamespaceProvider extends AdapterImpl implements ILibraryNamespaceProvider {
+
+    private final Map<String, Namespace> immutableNamespaces = new HashMap<>();
+
+    private final Notifier source;
+
+    public static LibraryNamespaceProvider get(ResourceSet rs) {
+        return rs.eAdapters().stream()
+                .filter(LibraryNamespaceProvider.class::isInstance)
+                .map(LibraryNamespaceProvider.class::cast)
+                .findFirst()
+                .orElse(null);
+    }
+
+    public static LibraryNamespaceProvider getFrom(EObject e) {
+        if (e != null) {
+            Resource resource = e.eResource();
+            if (resource != null) {
+                return getFrom(resource);
+            }
+        }
+        return null;
+    }
+
+    public static LibraryNamespaceProvider getFrom(Resource resource) {
+        ResourceSet resourceSet = resource.getResourceSet();
+        if (resourceSet != null) {
+            return get(resourceSet);
+        }
+        return null;
+    }
+
+    /**
+     * Simple constructor.
+     *
+     * @param source any notifier that is usable to compute the root {@link ResourceSet}.
+     */
+    public LibraryNamespaceProvider(Notifier source) {
+        super();
+        this.source = Objects.requireNonNull(source);
+    }
+
+    @Override
+    public Namespace getNamespaceFromLibrary(String qualifiedName) {
+        Namespace result = this.immutableNamespaces.get(qualifiedName);
+        if (result == null) {
+            result = this.resolve(qualifiedName);
+        }
+        return result;
+    }
+
+    @Override
+    public <T extends Namespace> T getNamespaceFromLibrary(String qualifiedName, Class<T> type) {
+        Namespace result = this.getNamespaceFromLibrary(qualifiedName);
+        if (type.isInstance(result)) {
+            return type.cast(result);
+        }
+        return null;
+    }
+
+    /**
+     * Adds all Namespaces of the {@link LibraryPackage} stored inside the root Namespace of the given Resource to this
+     * provider.
+     *
+     * @param resource
+     *            resource containing {@link LibraryPackage}
+     */
+    public void addRootImmutableLibraries(Resource resource) {
```

Comment:
I think this could be renamed because with the latest version of the file, there are no reference to immutable elsewhere in this class.
A proposition: addRootImmutableLibraries => addLibraryNamespaces

Reply:
As discussed it will become addImmutableLibrariesNamespaces